src/circuit/gadget/ecc/chip.rs

```diff
@@ -0,0 +1,386 @@
+        {
+            let config: witness_scalar_fixed::short::Config<C> = (&config).into();
+            config.create_gate(meta);
+        }
+
+        // Create incomplete point addition gate
+        {
+            let config: add_incomplete::Config = (&config).into();
+            config.create_gate(meta);
+        }
+
+        // Create complete point addition gate
+        {
+            let add_config: add::Config = (&config).into();
+            add_config.create_gate(meta);
+        }
+
+        // Create fixed-base scalar mul gates that are used in both full-width
+        // and short multiplication.
+        {
+            let mul_fixed_config: mul_fixed::Config<C, { constants::NUM_WINDOWS }> =
+                (&config).into();
+            mul_fixed_config.create_gate(meta);
+        }
+
+        // Create gates that are only used in short fixed-base scalar mul.
+        {
+            let short_config: mul_fixed::short::Config<C, { constants::NUM_WINDOWS_SHORT }> =
+                (&config).into();
+            short_config.create_gate(meta);
+        }
+
+        // Create variable-base scalar mul gates
+        {
+            let mul_config: mul::Config<C> = (&config).into();
+            mul_config.create_gate(meta);
+        }
+
+        config
+    }
+}
+
+/// A full-width scalar used for fixed-base scalar multiplication.
+/// This is decomposed in chunks of `window_width` bits in little-endian order.
+/// For example, if `window_width` = 3, we will have [k_0, k_1, ..., k_n]
+/// where `scalar = k_0 + k_1 * (2^3) + ... + k_n * (2^3)^n` and each `k_i` is
+/// in the range [0..2^3).
+#[derive(Clone, Debug)]
+pub struct EccScalarFixed<C: CurveAffine> {
+    value: Option<C::Scalar>,
+    windows: ArrayVec<CellValue<C::Base>, { constants::NUM_WINDOWS }>,
+}
+
+/// A signed short scalar used for fixed-base scalar multiplication.
+/// This is decomposed in chunks of `window_width` bits in little-endian order.
+/// For example, if `window_width` = 3, we will have [k_0, k_1, ..., k_n]
+/// where `scalar = k_0 + k_1 * (2^3) + ... + k_n * (2^3)^n` and each `k_i` is
+/// in the range [0..2^3).
+#[derive(Clone, Debug)]
+pub struct EccScalarFixedShort<C: CurveAffine> {
+    magnitude: Option<C::Scalar>,
+    sign: CellValue<C::Base>,
+    windows: ArrayVec<CellValue<C::Base>, { constants::NUM_WINDOWS_SHORT }>,
+}
+
+impl<C: CurveAffine> EccInstructions<C> for EccChip<C> {
+    type ScalarFixed = EccScalarFixed<C>;
+    type ScalarFixedShort = EccScalarFixedShort<C>;
+    type ScalarVar = CellValue<C::Base>;
+    type Point = EccPoint<C>;
+    type X = CellValue<C::Base>;
+    type FixedPoints = OrchardFixedBasesFull<C>;
+    type FixedPointsShort = ValueCommitV<C>;
+
+    fn witness_scalar_var(
+        &self,
+        layouter: &mut impl Layouter<C::Base>,
+        value: Option<C::Base>,
+    ) -> Result<Self::ScalarVar, Error> {
+        layouter.assign_region(
+            || "Witness scalar var",
```

Comment:
```suggestion
            || "witness scalar for variable-base mul",
```
(for consistency with below).